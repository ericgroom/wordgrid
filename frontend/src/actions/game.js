export const REQUEST_CREATE_GAME = "REQUEST_CREATE_GAME";
export const GAME_CREATED = "GAME_CREATED";
export const GAME_STATE_UPDATE = "GAME_STATE_UPDATE";
export const JOIN_GAME = "JOIN_GAME";
export const LEAVE_GAME = "LEAVE_GAME";
/**
 * Used by saga, since sagas run after reducers, and reducers
 * can't emit additional actions, it is difficult to both check
 * if a word has been sent to the backend and update state, so the
 * saga handles everything and then emits `ADD_WORD`
 */
export const WORD_COMPLETED = "WORD_COMPLETED";
export const UPDATE_WORD = "UPDATE_WORD";
export const REQUEST_START_GAME = "REQUEST_START_GAME";
export const START_GAME = "START_GAME";
export const END_GAME = "END_GAME";
export const START_COUNTDOWN = "START_COUNTDOWN";
export const REJOINED = "REJOINED";
export const ADD_WORD = "ADD_WORD";

export const createGame = () => ({ type: REQUEST_CREATE_GAME });
export const gameCreated = id => ({ type: GAME_CREATED, id });
export const joinGame = id => ({ type: JOIN_GAME, id });
export const addWord = word => ({ type: ADD_WORD, word });
export const updateWord = word => ({ type: UPDATE_WORD, word });
export const completeWord = word => ({ type: WORD_COMPLETED, word });

export const startGame = () => ({ type: REQUEST_START_GAME });
export const startCountdown = duration => ({ type: START_COUNTDOWN, duration });
export const leaveGame = id => ({ type: LEAVE_GAME, id });
export const endGame = () => ({ type: END_GAME });
export const rejoined = durationRemaining => ({
  type: REJOINED,
  durationRemaining
});

export const updateGameState = payload => ({
  type: GAME_STATE_UPDATE,
  state: payload
});
