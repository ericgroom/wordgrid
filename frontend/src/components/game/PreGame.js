import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { Helmet } from "react-helmet";
import BigButton from "../styles/BigButton";
import List from "../styles/List";
import Timer from "../shared/Timer";
import CopyableLink from "../shared/CopyableLink";

const Wrapper = styled.div`
  text-align: center;
  max-width: 600px;
  margin: 1rem auto;
  padding: 1rem;

  button {
    margin-bottom: 1rem;
  }
  .empty {
    color: rgba(0, 0, 0, 0.7);
  }
`;

class PreGame extends React.Component {
  static propTypes = {
    startGame: PropTypes.func.isRequired,
    users: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number,
        nickname: PropTypes.string
      })
    ).isRequired
  };

  state = {
    clicked: false
  };

  handleButtonClick = () => {
    this.setState({ clicked: true });
    this.props.startGame();
  };
  render() {
    return (
      <Wrapper>
        <Helmet>
          <title>{this.props.countdown ? "Starting!" : "In Lobby"}</title>
        </Helmet>
        <h3>
          Game isn't started yet, invite some friends with the link below!
        </h3>
        <CopyableLink url={window.location.toString()} />
        <BigButton
          onClick={this.handleButtonClick}
          disabled={this.state.clicked || this.props.countdown}
        >
          {this.props.countdown ? (
            <Timer duration={this.props.countdownDuration}>
              {remaining => `Starting Game in ${remaining}...`}
            </Timer>
          ) : (
            "Start Game"
          )}
        </BigButton>
        <List>
          <h2>Connected Users</h2>
          {this.props.users.length > 0 ? (
            <ul>
              {this.props.users.map(user => (
                <li key={user.id}>{user.nickname}</li>
              ))}
            </ul>
          ) : (
            <p className="empty">There's nobody else here yet</p>
          )}
        </List>
      </Wrapper>
    );
  }
}

export default PreGame;
