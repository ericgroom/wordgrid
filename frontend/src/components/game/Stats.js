import React from "react";
import styled from "styled-components/macro";
import AnimatedNumber from "../shared/AnimatedNumber";

const Wrapper = styled.div`
  text-align: center;
  h2 {
    height: 1rem;
  }
`;

const Stats = props => (
  <Wrapper>
    <h2>
      Score: <AnimatedNumber num={props.score} />
    </h2>
  </Wrapper>
);

export default Stats;
