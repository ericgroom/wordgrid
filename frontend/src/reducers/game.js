import get from "lodash/get";
import {
  REQUEST_CREATE_GAME,
  GAME_CREATED,
  GAME_STATE_UPDATE,
  JOIN_GAME,
  LEAVE_GAME,
  UPDATE_WORD,
  START_GAME,
  START_COUNTDOWN,
  REJOINED,
  END_GAME,
  ADD_WORD
} from "../actions";

const initialState = {
  id: null,
  created: false,
  started: false,
  ended: false,
  countdown: false,
  countdownDuration: 0,
  duration: 0,
  joined: false,
  loading: false,
  exists: true,
  grid: null,
  wordsById: {},
  wordIds: [],
  nextWordId: 0,
  users: []
};

const updateWord = (state, action) => {
  const word = state.wordsById[action.word.id];
  const updatedWord = { ...word, ...action.word };
  const wordsById = {
    ...state.wordsById,
    [action.word.id]: updatedWord
  };
  return {
    ...state,
    wordsById
  };
};

const addWord = (state, action) => {
  const wordIds = [...state.wordIds, action.word.id];
  const wordsById = { ...state.wordsById, [action.word.id]: action.word };
  return { ...state, wordIds, wordsById, nextWordId: state.nextWordId + 1 };
};

export default (state = initialState, action) => {
  switch (action.type) {
    case REQUEST_CREATE_GAME:
      return { ...state, created: true };
    case GAME_CREATED:
      return { ...state, id: action.id };
    case GAME_STATE_UPDATE:
      return { ...state, ...action.state, loading: false };
    case JOIN_GAME:
      return { ...state, id: action.id, joined: true, loading: true };
    case LEAVE_GAME:
      return initialState;
    case UPDATE_WORD:
      return updateWord(state, action);
    case ADD_WORD:
      return addWord(state, action);
    case START_GAME:
      return { ...state, started: true };
    case START_COUNTDOWN:
      return { ...state, countdown: true, countdownDuration: action.duration };
    case REJOINED:
      return {
        ...state,
        duration: action.durationRemaining
      };
    case END_GAME:
      return { ...state, ended: true };
    default:
      return state;
  }
};

/**
 * Get the score for a particular user
 * @param {object} state redux store state
 * @param {number} userId userId to fetch score for
 */
export const getScoreOfUser = (state, userId) => {
  return state.users.length > 0
    ? state.users
      ? get(state.users.find(user => user.id === userId), "score", 0)
      : 0
    : 0;
};

/**
 * Get the current state of the game as a string.
 * @param {object} state state local to this reducer
 */
export const getGameState = state => {
  if (state.loading) return "loading";
  if (!state.exists) return "non-existant";
  if (state.ended) return "ended";
  if (!state.started) return "pregame";
  if (state.started) return "active";
  if (process.env.NODE_ENV !== "production")
    throw new Error("unaccounted case in getGameState");
  return "non-existant";
};

/**
 * Gets all words played by the user.
 * @param {object} state state local to this reducer
 */
export const getAllWords = state =>
  state.wordIds.map(id => state.wordsById[id]);
