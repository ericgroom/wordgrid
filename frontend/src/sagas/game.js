import { eventChannel } from "redux-saga";
import { take, race, all, call, put, select } from "redux-saga/effects";
import max from "lodash/max";
import find from "lodash/find";

import {
  updateGameState,
  REQUEST_CREATE_GAME,
  gameCreated,
  JOIN_GAME,
  LEAVE_GAME,
  updateWord,
  WORD_COMPLETED,
  userJoined,
  sentWord,
  REQUEST_START_GAME,
  startCountdown,
  rejoined,
  endGame
} from "../actions";
import { putFrom, awaitAuthIfNeeded } from "./index";

function* gameSocketFlow(socket) {
  while (true) {
    try {
      console.log("restarting gameSocketFlow");
      const action = yield take(JOIN_GAME);
      yield call(awaitAuthIfNeeded);
      socket.emit("join game", { id: action.id });
      const socketChannel = yield call(gameSocketChannel, socket);
      yield race([putFrom(socketChannel), gameActionListener(socket)]);
    } catch (e) {
      console.error(e);
    }
  }
}

/**
 * Creates socket.io instance and emits actions based on events
 */
function gameSocketChannel(socket) {
  return eventChannel(emit => {
    socket.on("state", updatedState => {
      let state = { ...updatedState };
      if (state.grid) state.grid = state.grid.split("");
      emit(updateGameState(state));
    });
    socket.on("not exists", () => emit(updateGameState({ exists: false })));
    socket.on("word", word => {
      console.log(word);
      emit(updateWord(word));
    });
    socket.on("user join", nickname => {
      console.log(`${nickname} joined`);
      emit(userJoined(nickname));
    });
    socket.on("countdown", duration => {
      emit(startCountdown(duration));
    });
    socket.on("end game", () => {
      emit(endGame());
    });
    socket.on("remaining time", durationRemaining => {
      emit(rejoined(durationRemaining));
    });
    socket.on("played words", words => {
      emit(
        updateGameState({
          words,
          wordId: max(words.map(word => word.id)) + 1 || 0,
          sentWords: words.map(word => word.word)
        })
      );
    });
    return () => {
      emit({ type: LEAVE_GAME });
    };
  });
}

function* gameActionListener(socket) {
  yield call(awaitAuthIfNeeded);
  loop: while (true) {
    console.log("gameActionListener waiting...");
    const action = yield take([WORD_COMPLETED, REQUEST_START_GAME, LEAVE_GAME]);
    const gameId = yield select(state => state.game.id);
    switch (action.type) {
      case WORD_COMPLETED:
        // eslint-disable-next-line
        const { word, path } = action.word;
        // validate word
        // don't send words less than 3 letters long
        const tooShort = word.length < 3;
        // check if word has already been sent
        const sentWords = yield select(state => state.game.sentWords);
        const alreadySent = sentWords.some(sentWord => sentWord === word);

        // send the word
        if (!tooShort && !alreadySent) {
          const words = yield select(state => state.game.words);
          const wordId = find(words, { word }).id;
          const gameId = yield select(state => state.game.id);
          socket.emit("word", { word, wordId, gameId, path });
          yield put(sentWord(word));
        }
        break;
      case REQUEST_START_GAME:
        console.log(`starting game: ${gameId}`);
        socket.emit("game start", { id: gameId });
        break;
      case LEAVE_GAME:
        socket.emit("leave game", action.id);
        break loop;
      default:
        throw new Error(
          `messageActionListener saga took action but has no handler for: ${
            action.type
          }`
        );
    }
  }
}

/**
 * Listens to non-socket related game actions
 */
function* gameCreateListener() {
  while (true) {
    console.log("starting gameCreateListener");
    const action = yield take(REQUEST_CREATE_GAME);
    switch (action.type) {
      case REQUEST_CREATE_GAME:
        yield call(createGame);
        continue;
      default:
        throw new Error(
          `messageActionListener saga took action but has no handler for: ${
            action.type
          }`
        );
    }
  }
}

/**
 * Flow for creating a game.
 */
function* createGame() {
  try {
    const url = `${process.env.REACT_APP_BACKEND_URL}/game/new`;
    console.log(url);
    const res = yield call(fetch, url);
    const { gameId } = yield call([res, "json"]);
    yield put(gameCreated(gameId));
  } catch (e) {
    console.error(e);
  }
}

export default function* gameFlow(socket) {
  while (true) {
    yield all([gameSocketFlow(socket), gameCreateListener()]);
  }
}
