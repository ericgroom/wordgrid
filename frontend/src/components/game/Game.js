import React from "react";
import styled from "styled-components/macro";
import { Helmet } from "react-helmet";
import Timer from "../shared/Timer";
import WordGrid from "../shared/wordgrid";
import Stats from "./Stats";
import Scoreboard from "./Scoreboard";
import WordBank from "./WordBank";

const Container = styled.div`
  display: grid;
  gap: 1rem;
  align-items: center;
  justify-items: center;
  text-align: center;
  height: 100%;

  @media (min-width: 800px) {
    grid-template-columns: repeat(2, minmax(400px, 1fr));
  }

  @media (min-width: 1200px) {
    grid-template-columns: repeat(3, minmax(400px, 1fr));
    .scoreboard {
      order: 0;
    }
    .main {
      order: 1;
    }
    .wordbank {
      order: 2;
    }
  }
`;

class Game extends React.Component {
  render() {
    const {
      gameDuration,
      letters,
      onWordCompleted,
      score,
      connectedUsers,
      wordsPlayed
    } = this.props;
    return (
      <Container>
        <div className="main">
          <p className="mobile-warning">
            This website may not work on your mobile device.
          </p>
          <Timer duration={gameDuration}>
            {remaining => (
              <>
                <h1>{remaining}</h1>
                <Helmet>
                  <title>{`${remaining}`}</title>
                </Helmet>
              </>
            )}
          </Timer>
          {letters && <WordGrid letters={letters} onWord={onWordCompleted} />}
          <Stats score={score} />
        </div>
        <Scoreboard users={connectedUsers} className="scoreboard" />
        <aside className="wordbank">
          <WordBank words={wordsPlayed} />
        </aside>
      </Container>
    );
  }
}

export default Game;
