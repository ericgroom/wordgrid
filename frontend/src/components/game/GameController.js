import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { PoseGroup } from "react-pose";
import { joinGame, completeWord, startGame, leaveGame } from "../../actions";
import { getScoreOfCurrentUser, getGameState } from "../../reducers";
import PreGame from "./PreGame";
import Game from "./Game";
import PostGame from "./PostGame";
import Spinner from "../styles/Spinner";
import ToTheLeft from "../styles/ToTheLeft";
import Chat from "../chat";

class GameController extends Component {
  componentDidMount() {
    const id = this.props.match.params.id;
    if (id) {
      this.props.joinGame(id);
    } else {
      this.props.history.push("/");
    }
  }
  componentWillUnmount() {
    this.props.leaveGame(this.props.gameId);
  }

  render() {
    const { gameState } = this.props;
    switch (gameState) {
      case "loading":
        return <Spinner />;
      case "non-existant":
        return <h1>This game doesn't exist</h1>;
      case "active":
      case "pregame":
      case "ended":
        return (
          <div className="game">
            <PoseGroup>
              {gameState === "pregame" && (
                <ToTheLeft key={1}>
                  <PreGame {...this.props} users={this.props.connectedUsers} />
                </ToTheLeft>
              )}
              {gameState === "active" && (
                <ToTheLeft key={2}>
                  <Game {...this.props} />
                </ToTheLeft>
              )}
              {gameState === "ended" && (
                <ToTheLeft key={3}>
                  <PostGame results={this.props.connectedUsers} />
                </ToTheLeft>
              )}
            </PoseGroup>
            <Chat />
          </div>
        );
      default:
        throw new Error("unknown game state passed to <GameController />");
    }
  }
}

const mapStateToProps = state => {
  const {
    game,
    user: { nickname }
  } = state;
  return {
    letters: game.grid,
    gameId: game.id,
    gameState: getGameState(state),
    countdown: game.countdown,
    countdownDuration: game.countdownDuration,
    wordsPlayed: game.words,
    nickname: nickname,
    connectedUsers: game.users,
    gameDuration: game.duration,
    score: getScoreOfCurrentUser(state)
  };
};

const mapDispatchToProps = {
  joinGame,
  onWordCompleted: completeWord,
  startGame,
  leaveGame
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(GameController));
