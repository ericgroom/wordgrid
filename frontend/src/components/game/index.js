import GameController from "./GameController";
export default GameController;
