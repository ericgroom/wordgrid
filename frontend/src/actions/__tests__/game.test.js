import * as gameActions from "../game";

describe("game action creators", () => {
  it("createGame", () => {
    expect(gameActions.createGame()).toMatchSnapshot();
  });
  it("gameCreated", () => {
    expect(gameActions.gameCreated(123)).toMatchSnapshot();
  });
  it("joinGame", () => {
    expect(gameActions.joinGame(123)).toMatchSnapshot();
  });
  it("addWord", () => {
    expect(
      gameActions.addWord({
        id: 0,
        word: "asdf",
        path: [0, 1, 2, 3],
        valid: true
      })
    ).toMatchSnapshot();
  });
  it("updateWord", () => {
    expect(
      gameActions.updateWord({ id: 2, valid: false, score: 34 })
    ).toMatchSnapshot();
  });
  it("completeWord", () => {
    expect(
      gameActions.completeWord({ word: "asdf", path: [1, 2, 3, 4] })
    ).toMatchSnapshot();
  });
  /** doesn't take an id? */
  it("startGame", () => {
    expect(gameActions.startGame()).toMatchSnapshot();
  });
  it("startCountdown", () => {
    expect(gameActions.startCountdown(3)).toMatchSnapshot();
  });
  it("leaveGame", () => {
    expect(gameActions.leaveGame(123)).toMatchSnapshot();
  });
  /** no id? */
  it("endGame", () => {
    expect(gameActions.endGame()).toMatchSnapshot();
  });
  it("rejoined", () => {
    expect(gameActions.rejoined(15)).toMatchSnapshot();
  });
  it("updateGameState", () => {
    expect(
      gameActions.updateGameState({
        started: true,
        users: [{ id: 2, nickname: "Eric" }]
      })
    ).toMatchSnapshot();
  });
});
