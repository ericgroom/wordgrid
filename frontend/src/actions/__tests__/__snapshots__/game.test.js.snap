// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game action creators addWord 1`] = `
Object {
  "type": "ADD_WORD",
  "word": Object {
    "id": 0,
    "path": Array [
      0,
      1,
      2,
      3,
    ],
    "valid": true,
    "word": "asdf",
  },
}
`;

exports[`game action creators completeWord 1`] = `
Object {
  "type": "WORD_COMPLETED",
  "word": Object {
    "path": Array [
      1,
      2,
      3,
      4,
    ],
    "word": "asdf",
  },
}
`;

exports[`game action creators createGame 1`] = `
Object {
  "type": "REQUEST_CREATE_GAME",
}
`;

exports[`game action creators endGame 1`] = `
Object {
  "type": "END_GAME",
}
`;

exports[`game action creators gameCreated 1`] = `
Object {
  "id": 123,
  "type": "GAME_CREATED",
}
`;

exports[`game action creators joinGame 1`] = `
Object {
  "id": 123,
  "type": "JOIN_GAME",
}
`;

exports[`game action creators leaveGame 1`] = `
Object {
  "id": 123,
  "type": "LEAVE_GAME",
}
`;

exports[`game action creators rejoined 1`] = `
Object {
  "durationRemaining": 15,
  "type": "REJOINED",
}
`;

exports[`game action creators startCountdown 1`] = `
Object {
  "duration": 3,
  "type": "START_COUNTDOWN",
}
`;

exports[`game action creators startGame 1`] = `
Object {
  "type": "REQUEST_START_GAME",
}
`;

exports[`game action creators updateGameState 1`] = `
Object {
  "state": Object {
    "started": true,
    "users": Array [
      Object {
        "id": 2,
        "nickname": "Eric",
      },
    ],
  },
  "type": "GAME_STATE_UPDATE",
}
`;

exports[`game action creators updateWord 1`] = `
Object {
  "type": "UPDATE_WORD",
  "word": Object {
    "id": 2,
    "score": 34,
    "valid": false,
  },
}
`;
